src/utils.js

```diff
@@ -224,6 +226,44 @@ export function mergeObjects(obj1, obj2, concatArrays = false) {
   }, acc);
 }
 
+function objectNeedsTrimmed(obj) {
+  // Recursively determine if an object needs to be simplified to `undefined`.
+  return Object.keys(obj).every(key => {
+    if (obj.hasOwnProperty(key) && isObject(obj[key])) {
+      return objectNeedsTrimmed(obj[key]);
+    }
+    return typeof obj[key] === "undefined";
+  });
+}
+
+export function trimObject(obj) {
+  // Recursively trim an object converting objects whose values are all undefined to `undefined`.
+  let result = Object.assign({}, obj); // Prevent mutation of source object.
+  for (const key of Object.keys(result)) {
+    if (result.hasOwnProperty(key) && isObject(result[key])) {
+      result[key] = trimObject(result[key]);
+    }
+  }
```

Comment:
We may want to benchmark this, but I wonder how much slower would be to just use `JSON.parse(JSON.stringify(obj))`. I've read that these fns were heavily optimized in modern js engines. That would give us:

- Deep cloning for free,
- Getting rid of the `objectNeedsTrimmed` function call.

Reply:
I am not sure simply calling `JSON.parse(JSON.stringify(obj))` would work I  tried it and it converts `{"a": true, "b": {"c": undefined, "d": {"e": undefined}}}` to `{ "a": true, "b": {"d": {}}}` and I believe we would need it to convert it to `{ "a": true, "b":undefined}`. That being said we could now simply remove empty objects and get the same behavior, but we still have to walk through the fields of the object recursively.